### PR TITLE
[Ruby Client] Correct caching and usage of source vs. "en" bundles

### PR DESCRIPTION
**Pull request by jessiejuachon**
Part 1 of "splitting [pull 1550](https://github.com/vmware/singleton/pull/1550) into 3 parts":

1.  Loading source bundle in cache and using them directly when target locale is same as source locale
2. Source comparison
3. Moving some methods from translation.rb into a new base.rb file.

### CODE REVIEW

**Review by Xiaochao8 — DISMISSED** [review]

**Review by deminyan — DISMISSED** [review]

**Review by deminyan — DISMISSED** [review]

File: lib/sgtn-client/util/locale-util.rb [hunk 1]
```diff
@@ -10,6 +12,11 @@ module SgtnClient
                   }
 
     class LocaleUtil
+        def self.get_best_locale(locale)
+            flocale = fallback(process_locale(locale))
+            flocale = SgtnClient::Config.configurations.default if flocale == get_source_locale()
+            return flocale
```
Comment (Xiaochao8) [inline lib/sgtn-client/util/locale-util.rb:18]: Change 'en' to 'default' will cause a request to server with locale 'default'.
```suggestion
            fallback(process_locale(locale))
```

File: lib/sgtn-client/api/translation.rb [hunk 2]
```diff
@@ -56,6 +58,7 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:61]: If locale is source locale, return source directly.
```suggestion
        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
        return getSources(component, locale) if locale == LocaleUtil.get_source_locale

```

File: lib/sgtn-client/api/translation.rb [hunk 3]
```diff
@@ -77,6 +80,7 @@ def self.getStrings(component, locale)
       private
 
       def self.getTranslation(component, key, locale)
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
         items = get_cs(component, locale)
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:84]: If locale is source locale, return source directly.
```suggestion
        if locale == LocaleUtil.get_source_locale
          items = getSources(component, locale)
        else
          items = get_cs(component, locale)
        end
```

**Review by Xiaochao8 — CHANGES_REQUESTED** [review]

File: lib/sgtn-client/util/locale-util.rb (hunk 1 above)
Comment (jessiejuachon) [reply lib/sgtn-client/util/locale-util.rb:18]: yes, this will cause it to use the source bundle

File: lib/sgtn-client/api/translation.rb (hunk 3 above)
Comment (jessiejuachon) [reply lib/sgtn-client/api/translation.rb:84]: get_best_locale will replace 'en' with SgtnClient::Config.configurations.default so the source bundle will be loaded.

File: lib/sgtn-client/api/translation.rb (hunk 2 above)
Comment (jessiejuachon) [reply lib/sgtn-client/api/translation.rb:61]: get_best_locale will replace 'en' with SgtnClient::Config.configurations.default so the source bundle will be loaded.

File: lib/sgtn-client/util/locale-util.rb (hunk 1 above)
Comment (Xiaochao8) [reply lib/sgtn-client/util/locale-util.rb:18]: Check server log, received a strange request.
```
2022-03-09T02:04:13.400Z        INFO    Start a request {“traceId”: 1120116155, “method”: “GET”, “ip”: “127.0.0.1”, “path”: “/i18n/api/v2/translation/products/test/versions/4.8.1/locales/default/components/i18n”, “query”: “checkTranslationStatus=false&machineTranslation=false&pseudo=false”}
2022-03-09T02:04:13.407Z        DEBUG   Get a bundle    {“traceId”: 1120116155, “name”: “test”, “version”: “4.8.1", “locale”: “default”, “component”: “i18n”}
2022-03-09T02:04:13.408Z        DEBUG   Read a bundle from storage      {“traceId”: 1120116155, “name”: “test”, “version”: “4.8.1”, “locale”: “default”, “component”: “i18n”}
2022-03-09T02:04:13.413Z        ERROR   Fail to get translation for test/4.8.1/default/i18n: open tests/testdata/bundles/test/4.8.1/i18n/messages_default.json: no such file or directory       {“traceId”: 1120116155}
```

**Review by Xiaochao8 — DISMISSED** [review]

**Review by linr211 — APPROVED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 4]
```diff
@@ -56,6 +58,9 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        return getSources(component, locale) if locale == LocaleUtil.get_source_locale
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:62]:
```suggestion
        return Source.getSources(component, locale) if locale == LocaleUtil.get_source_locale
```

**Review by Xiaochao8 — APPROVED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 5]
```diff
@@ -77,7 +82,12 @@ def self.getStrings(component, locale)
       private
 
       def self.getTranslation(component, key, locale)
-        items = get_cs(component, locale)
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        if locale == LocaleUtil.get_source_locale
+          items = Source.getSources(component, locale)
+        else
+          items = get_cs(component, locale)
+        end
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:90]:
```suggestion
        items = get_cs(component, locale)
```

File: lib/sgtn-client/api/translation.rb [hunk 6]
```diff
@@ -56,6 +58,9 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        return Source.getSources(component, locale) if locale == LocaleUtil.get_source_locale
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:62]:
```suggestion
        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
```

File: lib/sgtn-client/api/translation.rb [hunk 7]
```diff
@@ -86,13 +96,11 @@ def self.getTranslation(component, key, locale)
       end
 
       def self.get_cs(component, locale)
-        locale = SgtnClient::LocaleUtil.process_locale(locale)
-        flocale = SgtnClient::LocaleUtil.fallback(locale)
-        cache_key = SgtnClient::CacheUtil.get_cachekey(component, flocale)
+        cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:99]:
```suggestion
        # source locale always return source bundle
        if locale == LocaleUtil.get_source_locale
          sources = SgtnClient::Source.getSources(component, SgtnClient::Config.configurations.default)
          messages = sources&.first&.last
          return {'locale' => locale, 'component' => component, 'messages' => messages} if messages
        end

        cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
```

File: lib/sgtn-client/api/translation.rb (hunk 5 above)
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:90]:
```suggestion
        items = get_cs(component, locale)
```

File: lib/sgtn-client/api/translation.rb [hunk 8]
```diff
@@ -56,6 +58,9 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
+        return Source.getSources(component, locale) if locale == LocaleUtil.get_source_locale
+
```
Comment (Xiaochao8) [inline lib/sgtn-client/api/translation.rb:63]:
```suggestion

```

**Review by Xiaochao8 — APPROVED** [review]

**Review by Xiaochao8 — APPROVED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 9]
```diff
@@ -59,6 +59,7 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
         items = get_cs(component, locale)
```
Comment (linr211) [inline lib/sgtn-client/api/translation.rb:63]:
```suggestion
        items = get_cs(component, locale) if locale == LocaleUtil.get_source_locale
```

File: lib/sgtn-client/api/translation.rb [hunk 10]
```diff
@@ -89,13 +91,18 @@ def self.getTranslation(component, key, locale)
...
-        locale = SgtnClient::LocaleUtil.process_locale(locale)
-        flocale = SgtnClient::LocaleUtil.fallback(locale)
-        cache_key = SgtnClient::CacheUtil.get_cachekey(component, flocale)
+        # source locale always return source bundle
+        if locale == LocaleUtil.get_source_locale
+          sources = SgtnClient::Source.getSources(component, SgtnClient::Config.configurations.default)
+          messages = sources&.first&.last
+          return {'locale' => locale, 'component' => component, 'messages' => messages} if messages
+        end
+
```
Comment (linr211) [inline lib/sgtn-client/api/translation.rb:100]: It's duplicated with similar codes in getStrings
```suggestion

```

File: lib/sgtn-client/api/translation.rb [hunk 11]
```diff
@@ -59,6 +59,7 @@ def self.getString_f(component, key, args, locale, *optionals)
 
       def self.getStrings(component, locale)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
         items = get_cs(component, locale)
         default = SgtnClient::Config.configurations.default
         if items.nil? || items["messages"] == nil
```
Comment (linr211) [inline lib/sgtn-client/api/translation.rb:65]:
```suggestion
        items = get_cs(component, locale) if locale != LocaleUtil.get_source_locale
        default = SgtnClient::Config.configurations.default
        if (defined?(items)).nil? || items.nil? || items["messages"] == nil
```

File: lib/sgtn-client/api/translation.rb [hunk 12]
```diff
@@ -80,6 +81,7 @@ def self.getStrings(component, locale)
       private
 
       def self.getTranslation(component, key, locale)
+        locale = SgtnClient::LocaleUtil.get_best_locale(locale)
         items = get_cs(component, locale)
```
Comment (linr211) [inline lib/sgtn-client/api/translation.rb:85]:
```suggestion
        items = getStrings(component, locale)
```

File: lib/sgtn-client/api/translation.rb (hunk 10 above)
Comment (Xiaochao8) [reply lib/sgtn-client/api/translation.rb:100]: Will this introduct a problem? If not, I prefer to perceed with current to make changes as small as possible.

File: lib/sgtn-client/api/translation.rb (hunk 11 above)
Comment (Xiaochao8) [reply lib/sgtn-client/api/translation.rb:65]: Looks complex. If there isn't a bug, I prefer to perceed with current to make changes as small as possible.

File: lib/sgtn-client/api/translation.rb (hunk 12 above)
Comment (Xiaochao8) [reply lib/sgtn-client/api/translation.rb:85]: Looks both work. This PR is to fix bug, let's not change this if there isn't a bug.

File: lib/sgtn-client/api/translation.rb (hunk 10 above)
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:100]: suggested change is smaller actually, i approve it to speed up the progress.

**Review by linr211 — APPROVED** [review]

**Pull request merged**